### ISSUE TITLE
Reduce Duplication Between SatisfyingAssignment and WitnessCS Structs

### ISSUE THREAD

**Issue by huitseeker** [lurk-lab/arecibo#77]
It has been observed that the `SatisfyingAssignment` (in this repo) and `WitnessCS` (in [bellpepper](https://github.com/lurk-lab/bellpepper)) structs have significant overlap in their definitions. This duplication can lead to maintenance challenges and potential inconsistencies. 

#### Similarities:
1. **Purpose**: Both structs serve as a `ConstraintSystem` that calculates witness values for a concrete instance of an R1CS circuit.
2. **Fields**: Both structs have fields `input_assignment` and `aux_assignment`.
3. **Methods**: Most methods in both structs are identical in both functionality and structure. This includes the implementation of the `ConstraintSystem` trait, as well as auxiliary methods such as `scalar_inputs` and `scalar_aux`.

#### Differences:
1. **Generics**:
    - `SatisfyingAssignment` is parameterized over `G: Group` and uses `G::Scalar`.
    - `WitnessCS` is parameterized over `Scalar` where `Scalar: PrimeField`.
2. **Attributes & Derivations**:
    - `WitnessCS` has the `#[derive(Debug, Clone, PartialEq, Eq)]` attribute, which provides automatic implementations of the Debug, Clone, PartialEq, and Eq traits.
    - `SatisfyingAssignment` provides its own `fmt::Debug` and `PartialEq` implementations and does not derive Clone or Eq.

#### Refactoring Potential:

1. **Shared Trait or Base Struct**: Given the similarity in behavior, a shared trait or a base struct could be created to capture the common functionalities. The specific implementations (`SatisfyingAssignment` and `WitnessCS`) can then extend or implement this shared definition.
2. **Consolidate Attributes & Derivations**: Determine whether `SatisfyingAssignment` requires its own Debug and PartialEq implementations. If not, it could use the derived implementations similar to `WitnessCS`.
3. **Type Alias or Adaptors**: If the core difference lies in the type parameters (`G::Scalar` vs. `Scalar`), consider type aliasing or using adaptors to harmonize the two.

**Comment by jobez** [comment]
i can take this!

**Comment by jobez** [comment]
Ok, got into looking at this today! 

Given that:

-  Group::Scalar should be able to be seen as satisfying PrimeField by the compiler, given the appropriate type annotations and constructor hints

https://github.com/lurk-lab/arecibo/blob/170818e04d76f0f4c71ca797c62db4782cfae8e2/src/traits/mod.rs#L37-L43

- The custom implementations for Debug and PartialEq were found to be unnecessary as the default implementations provided the required functionality, as verified by running the following commands with updated code:
    - cargo test --release
    - cargo run --release --example minroot
    - cargo nextest run --profile ci --workspace --cargo-profile dev-ci
- The implementation of `ConstraintSystem` across `SatisfyingAssignment` in arecibo and `WitnessCS` are the same

I see two available options, where the deciding factor is how likely is `SatisfyingAssignment` going to need to diverge from `WitnessCS` going forward.

## Newtype Pattern: Defining SatisfyingAssignment Struct

The most flexible or adaptable approach is to define `SatisfyingAssignment` like so:

```
pub struct SatisfyingAssignment<G: traits::Group>(WitnessCS<<G as Group>::Scalar>);
```

We would implement `ConstraintSystem` on our new type and simply call into the WitnessCS implementation. Further, we would expose `aux_assignment` and `input_assignment` as methods that would call into the wrapped WitnessCS instance. This both consolidates behavior and gives future developers the ability to differentiate if they so need.

## Type Alias approach

The most straightforward way to consolidate shared behavior is to *type alias* SatisfyingAssignment in `arecibo`, ala

```
pub type SatisfyingAssignment<G> = WitnessCS<<G as Group>::Scalar>;
```

Unfortunately this 'direct' consolidation comes with it a small bit of a refactor where calls like this

```
let mut cs: SatisfyingAssignment<G> = SatisfyingAssignment::new();
```

turn into calls like this:

```
let mut cs: SatisfyingAssignment<G> = SatisfyingAssignment::<G>::new();
```

This occurs because the type alias introduces an intermediate step, requiring a bit more complexity for the programmer, as they have to be more explicit about the generic type. For example, specifying <G> helps the compiler understand the exact type being referred to.


In both cases, we would also need to do a small refactor in bellpepper where we make the WitnessCS struct's fields public, outside of just the crate.

I prefer the first approach, but perhaps someone more familiar with the codebase would think the second would be better suited.

Pending feedback, I can get started on a PR!

**Comment by huitseeker** [comment]
- The Lurk-lab organization owns both bellpepper and lurk-rs. As a result, making PRs to bellpepper is acceptable.
- The Type Alias approach is preferred as it provides more simplicity, eliminating the need to expose newtype methods.
- For the compiler, only one type of information is required, which can be either:
```rust
let mut cs = SatisfyingAssignment::<G>::new();
```
OR
```rust
let mut cs: SatisfyingAssignment<G> = SatisfyingAssignment::new();
```

**Comment by huitseeker** [comment]
That seems fine!

**Comment by huitseeker** [comment]
Closed in #89
